File: ACM/2.字母异位词分组.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine().trim());
        String[] strs = br.readLine().split("\\s+");
        List<List<String>> result = new Solution().groupAnagrams(strs);
        for (List<String> group : result) {
            System.out.println(String.join(" ", group));
        }
    }
}

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap();
        for (String str : strs) {
            char[] chars = str.toCharArray();
            Arrays.sort(chars);
            String key = new String(chars);
            if (!map.containsKey(key)) {
                map.put(key, new ArrayList<>());
            }
            map.get(key).add(str);
        }
        return new ArrayList(map.values());
    }
}

File: ACM/2.字母异位词分组.py
import sys
from collections import defaultdict
def group_anagrams(strs):
    groups = defaultdict(list)
    for s in strs:
        key = ''.join(sorted(s))
        groups[key].append(s)
    return list(groups.values())

if __name__ == "__main__":
    # 读取输入，第一行是整数n，第二行是n个字符串，以空格分隔
    n = int(sys.stdin.readline().strip())
    strs = sys.stdin.readline().split()
    
    result = group_anagrams(strs)
    for group in result:
        print(' '.join(group))